master/internal/api_generic_tasks.go

```diff
@@ -0,0 +1,316 @@
+	}
+
+	launchWarnings, err := a.m.rm.ValidateResourcePoolAvailability(
+		&sproto.ValidateResourcePoolAvailabilityRequest{
+			Name:  poolName,
+			Slots: *resources.SlotsPerTask,
+		},
+	)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("checking resource availability: %v", err.Error())
+	}
+	if a.m.config.ResourceManager.AgentRM != nil &&
+		a.m.config.LaunchError &&
+		len(launchWarnings) > 0 {
+		return nil, nil, nil, fmt.Errorf("slots requested exceeds cluster capacity")
+	}
+
+	// Get the base TaskSpec.
+	taskContainerDefaults, err := a.m.rm.TaskContainerDefaults(
+		poolName,
+		a.m.config.TaskContainerDefaults,
+	)
+	if err != nil {
+		return nil, nil, nil, fmt.Errorf("getting TaskContainerDefaults: %v", err)
+	}
+	taskSpec := *a.m.taskSpec
+	taskSpec.TaskContainerDefaults = taskContainerDefaults
+	taskSpec.AgentUserGroup = agentUserGroup
+	taskSpec.Owner = userModel
+
+	// Get the full configuration.
+	taskConfig := model.DefaultConfigGenericTaskConfig(&taskSpec.TaskContainerDefaults)
+	workDirInDefaults := taskConfig.WorkDir
+
+	if err := yaml.Unmarshal([]byte(configYAML), &taskConfig); err != nil {
+		return nil, nil, nil, fmt.Errorf("yaml unmarshaling generic task config: %w", err)
+	}
+
+	// Copy discovered (default) resource pool name and slot count.
+	taskConfig.Resources.RawResourcePool = &poolName
+	taskConfig.Resources.RawSlotsPerTask = resources.SlotsPerTask
+
+	taskContainerPodSpec := taskSpec.TaskContainerDefaults.GPUPodSpec
+	if taskConfig.Resources.SlotsPerTask() == 0 {
+		taskContainerPodSpec = taskSpec.TaskContainerDefaults.CPUPodSpec
+	}
+	taskConfig.Environment.PodSpec = (*k8sV1.Pod)(schemas.Merge(
+		(*expconf.PodSpec)(taskConfig.Environment.PodSpec),
+		(*expconf.PodSpec)(taskContainerPodSpec),
+	))
+
+	var contextDirectoryBytes []byte
+	if len(contextDirectory) > 0 {
+		userFiles := filesToArchive(contextDirectory)
+
+		workdirSetInReq := taskConfig.WorkDir != nil &&
+			(workDirInDefaults == nil || *workDirInDefaults != *taskConfig.WorkDir)
+		if workdirSetInReq {
+			return nil, nil, nil, status.Errorf(codes.InvalidArgument,
+				"cannot set work_dir and context directory at the same time")
+		}
+		taskConfig.WorkDir = nil
+
+		contextDirectoryBytes, err = archive.ToTarGz(userFiles)
+		if err != nil {
+			return nil, nil, nil, status.Errorf(codes.InvalidArgument,
+				fmt.Errorf("compressing files context files: %w", err).Error())
+		}
+	}
+
+	extConfig := config.GetMasterConfig().InternalConfig.ExternalSessions
+	var token string
+	if extConfig.Enabled() {
+		token, err = grpcutil.GetUserExternalToken(ctx)
+		if err != nil {
+			return nil, nil, nil, status.Errorf(codes.Internal,
+				errors.Wrapf(err,
+					"unable to get external user token").Error())
+		}
+	} else {
+		token, err = user.StartSession(ctx, userModel)
+		if err != nil {
+			return nil, nil, nil, status.Errorf(codes.Internal,
+				errors.Wrapf(err,
+					"unable to create user session inside task").Error())
+		}
+	}
```

Comment:
I really think we should refactor this so it's not copied from command/experiment task specs, but happens in a single place in the code. we had issues with task specs so many times when we did something in one of them but not in another. if we'd have three of them, it'd become even riskier.